### PR TITLE
[exporter/clickhouse] Default `async_insert` to true. Added related config option.

### PR DESCRIPTION
**Pull request by SpencerTorres**
### DEPENDS ON #33693, #33694

**Description:**
Sets `async_insert` to true by default to enable [asynchronous inserts](https://clickhouse.com/docs/en/optimize/asynchronous-inserts).
Because this value is being given a default, I have added a config option under the same name.
Keep in mind that if `async_insert` is provided in `endpoint` or `connection_params` it will take precedence and ignore this new config option.

This is similar to how the `database` config option behaves.
The goal is to provide better insert performance by default, since not all users will know to set it in their DSN URL.

This also opens the discussion to ___**whether or not this is a breaking change**___. Depending on the deployment's telemetry throughput, this could be an unexpected change that leads to [`TOO_MANY_PARTS`](https://clickhouse.com/docs/knowledgebase/exception-too-many-parts) errors. I don't expect this to be the case however, but I welcome any discussion about this concern.

This PR is being resubmitted with suggestions from @crobert-1 and @dmitryax  applied.
Here are the extra changes with these suggestions applied:
- Extracted unrelated changes into separate PRs
- Updated `async_insert` to avoid using a `bool` pointer
- Updated tests to be able to support these non-pointer-yet-still-optional test cases

**Testing:**
Ran integration tests. Also added an abundance of tests to check the behavior of `async_insert` when present in `endpoint`, `connection_params`, and exporter config.

**Documentation:**
- Updated README for all related changes

Unrelated change, also updated README's SQL samples to use `sql` instead of `clickhouse` for the code samples to enable proper syntax highlighting. ClickHouse SQL is compatible with plain SQL.

### CODE REVIEW

**Comment by crobert-1** [comment]
Note for future reference, this is a continuation of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32340

File: exporter/clickhouseexporter/README.md [hunk 1]
```diff
@@ -282,7 +282,9 @@ Connection options:
 - `connection_params` (default = {}). Params is the extra connection parameters with map format.
 - `ttl_days` (default = 0): **Deprecated: Use 'ttl' instead.**  The data time-to-live in days, 0 means no ttl.
 - `ttl` (default = 0): The data time-to-live example 30m, 48h. Also, 0 means no ttl.
-- `database` (default = otel): The database name.
+- `database` (default = default): The database name. Overrides the database defined in `endpoint` when this setting is not equal to `default`.
+- `connection_params` (default = {}). Params is the extra connection parameters with map format. Query parameters provided in `endpoint` will be individually overwritten if present in this map.
```
Comment (dmitryax) [inline exporter/clickhouseexporter/README.md:286]: `connection_params` is already defined above

File: exporter/clickhouseexporter/config.go [hunk 2]
```diff
@@ -151,11 +150,7 @@ func (cfg *Config) buildDB(database string) (*sql.DB, error) {
 
 // ShouldCreateSchema returns true if the exporter should run the DDL for creating database/tables.
 func (cfg *Config) ShouldCreateSchema() bool {
```
Comment (dmitryax) [inline exporter/clickhouseexporter/config.go:152]: There is no reason to have these methods exported. All all them should be private.

Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:152]: Submitted https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33647

Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:152]: Then, this one can be actually removed

Comment (SpencerTorres) [reply exporter/clickhouseexporter/config.go:152]: Agreed, we can merge this and then update the other PR?

**Comment by SpencerTorres** [comment]
Looks like the README got merged incorrectly, I've removed the duplicate 👍

File: exporter/clickhouseexporter/config.go (hunk 2 above)
Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:152]: They can be merged independently

File: exporter/clickhouseexporter/config.go [hunk 3]
```diff
@@ -107,19 +111,14 @@ func (cfg *Config) buildDSN(database string) (string, error) {
...
 	}
 
-	// Override database if specified in config.
-	if cfg.Database != "" {
-		dsnURL.Path = cfg.Database
-	}
-
-	// Override database if specified in database param.
-	if database != "" {
-		dsnURL.Path = database
```
Comment (dmitryax) [inline exporter/clickhouseexporter/config.go:117]: Are these changes unrelated to this PR? If so, it should be another PR. Otherwise, the code review becomes confusing.

Ideally, even unrelated README updates should go to a separate PR.

Comment (SpencerTorres) [reply exporter/clickhouseexporter/config.go:117]: I agree, typically this would be a separate PR, but updating the tests affected a lot of this other config code. This PR goes beyond the title, but corrects a lot of these tests/config behaviors. I can rename the PR+changelog if that would be better.

Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:117]: Yeah, let's break it down into separate independent PRs if it's applicable. It'll be easy to review and track down changes in git history afterthey are merged

**Comment by SpencerTorres** [comment]
@dmitryax did some git surgery to hopefully extract all these changes into separate PRs. The PRs do build on one another, so I've ordered them so that they can be merged without breaking main.

**Comment by dmitryax** [comment]
@hanjm @Frapschen PTAL

**Review by dmitryax — APPROVED** [review]

**Review by hanjm — APPROVED** [review]
LGTM

**Comment by SpencerTorres** [comment]
Thanks for reviewing! I don't have permission to merge, but after this is merged we can do #33615

File: exporter/clickhouseexporter/factory.go [hunk 4]
```diff
@@ -38,12 +38,13 @@ func createDefaultConfig() component.Config {
 		QueueSettings:    queueSettings,
 		BackOffConfig:    configretry.NewDefaultBackOffConfig(),
 		ConnectionParams: map[string]string{},
-		Database:         defaultDatabase,
+		Database:         "default",
```
Comment (TylerHelmuth) [inline exporter/clickhouseexporter/factory.go:41]: The const `defaultDatabase` still exists in the code with a value of `"default"`, can we use that instead?

Comment (SpencerTorres) [reply exporter/clickhouseexporter/factory.go]: I originally avoided this because I didn't like using the const from `config.go`. I have applied this change for now just so we can get this merged asap.

**Comment by SpencerTorres** [comment]
Latest suggestions applied + rebased, hopefully we can get this merged asap since it's been open for a while. Thanks!

**Review by TylerHelmuth — APPROVED** [review]
